Refactor device state tracking in PointerEventConverter

dkwingsmt opened the pull request:
## Description
This PR refactors `_PointerState`.
- It is renamed to `_DeviceState` because its instances are indexed by device ID instead of pointer ID.
- `_DeviceState` now stores the entire last event, instead of `pointer` and `lastPosition`. Its meaning is much clearer after the change, tracking the state of a *device* at a given point of time, which can contain a pointer or might not.
- Some synthetic events that are not assigned with correct value have been fixed.
- Fixed an issue where `ui.PointerChange.remove` with position change is not expanded to any events with position change.
- Added several `assert`s to make sure the consumed events are to the same device, same pointer, and its `delta` correctly reflects the position difference from the last event.

## Related Issues

N/A

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

dkwingsmt committed:
Add "changes" to PointerEvents.

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into submenu

dkwingsmt committed:
Fix lint

dkwingsmt committed:
Fix lint

Hixie commented:
Generally we have been preferring level-based APIs to edge-based APIs, which is the reverse of the approach here. Is there a design doc or issue that discusses the design here so I can learn more about the trade-off being considered?

Hixie commented on packages/flutter/lib/src/gestures/converter.dart:
```diff
@@ -4,65 +4,144 @@
 import 'dart:ui' as ui show PointerData, PointerChange, PointerSignalKind;
 
-import 'package:flutter/foundation.dart' show visibleForTesting;
+import 'package:flutter/foundation.dart' show visibleForTesting, required;
```
just remove the `show`, it's not necessary for foundation.

Hixie commented:
(Please get an LGTM from me before checking this in; I'd like to make sure I understand the design and purpose before we add something like this as this is something it would be very hard to change back later.)

dkwingsmt commented:
@Hixie I will. Also I linked the related design doc in chat.

Hixie commented:
Cool, thanks.

dkwingsmt committed:
Remove PointerDifference and changes

dkwingsmt committed:
Merge branch 'pointer-event-changes' of github.com:dkwingsmt/flutter into pointer-event-changes

dkwingsmt commented:
@Hixie Can you take a look at the current PR? I removed the `changes` property, but I think the remaining changes are worth merging as a refactor. I've changed the PR description to include changes.

goderbauer commented:
Please also talk to @chunhtai to make sure this will work with his refactoring.

dkwingsmt commented:
This PR will be split into multiple parts. The first one is https://github.com/flutter/flutter/pull/29998.
